rlp/src/stream.rs

```diff
@@ -26,43 +27,39 @@ impl ListInfo {
 }
 
 /// Appendable rlp encoder.
-pub struct RlpStream {
+pub struct RlpStream<'a> {
 	unfinished_lists: Vec<ListInfo>,
-	buffer: Vec<u8>,
+	start_pos: usize,
+	buffer: &'a mut BytesMut,
 	finished_list: bool,
 }
 
-impl Default for RlpStream {
-	fn default() -> Self {
-		RlpStream::new()
-	}
-}
-
-impl RlpStream {
+impl<'a> RlpStream<'a> {
 	/// Initializes instance of empty `Stream`.
-	pub fn new() -> Self {
-		RlpStream { unfinished_lists: Vec::with_capacity(16), buffer: Vec::with_capacity(1024), finished_list: false }
+	pub fn new(buffer: &'a mut BytesMut) -> Self {
```

Comment:
I'm not a fan of tying this lib to only `BytesMut`, is it possible to make it generic so that it can accept both a vec and bytesmut?

Reply:
There is `BufMut` trait, but it is rather abridged in that it does not provide methods for controlling allocation and determining current buffer length. Besides, trying to make `RlpStream` generic would break `Encodable`'s object safety.

Reply:
Ok, I see.